includes/oauth/class-google-login.php

```diff
@@ -0,0 +1,182 @@
+		return Google_OAuth::api_google_auth_get_url(
+			[
+				'csrf_token'     => OAuth::generate_csrf_token( self::CSRF_TOKEN_NAMESPACE ),
+				'scope'          => implode( ' ', self::REQUIRED_SCOPES ),
+				'redirect_after' => add_query_arg( self::AUTH_CALLBACK, wp_create_nonce( self::AUTH_CALLBACK ), get_home_url() ),
+			]
+		);
+	}
+
+	/**
+	 * OAuth callback.
+	 */
+	public static function oauth_callback() {
+		if ( ! isset( $_GET[ self::AUTH_CALLBACK ] ) ) {
+			return;
+		}
+
+		if ( ! wp_verify_nonce( sanitize_text_field( $_GET[ self::AUTH_CALLBACK ] ), self::AUTH_CALLBACK ) ) {
+			wp_die( esc_html__( 'Invalid nonce.', 'newspack' ) );
+			return;
+		}
+
+		if ( ! isset( $_REQUEST['csrf_token'] ) || ! isset( $_REQUEST['access_token'] ) ) {
+			wp_die( esc_html__( 'Invalid request', 'newspack' ) );
+			return;
+		}
+
+		$saved_csrf_token = OAuth::retrieve_csrf_token( self::CSRF_TOKEN_NAMESPACE );
+
+		if ( $_REQUEST['csrf_token'] !== $saved_csrf_token ) {
+			Logger::log( 'Failed saving email - CSRF token mismatch.' );
+			\wp_die( \esc_html__( 'Authentication failed.', 'newspack' ) );
+		}
+
+		$user_email = Google_OAuth::validate_token_and_get_email_address( sanitize_text_field( $_REQUEST['access_token'] ), self::REQUIRED_SCOPES );
+		if ( is_wp_error( $user_email ) ) {
+			Logger::log( 'Failed validating user: ' . $user_email->get_error_message() );
+			\wp_die( \esc_html__( 'Authentication failed.', 'newspack' ) );
+		}
+
+		Logger::log( 'Got user email from Google: ' . $user_email );
+
+		// Associate the email address with the client ID for later retrieval.
+		$client_id = Reader_Activation::get_client_id();
+		set_transient( self::EMAIL_TRANSIENT_PREFIX . $client_id, $user_email, 20 );
+
+		/** Close window if it's a popup. */
+		?>
+		<script type="text/javascript" data-amp-plus-allowed>
+			if ( window.opener ) { window.close(); }
+		</script>
+		<?php
+	}
+
+	/**
+	 * Get Google authentication status.
+	 */
+	public static function api_google_login_register() {
+		// Get unique identifier of the client.
+		$client_id = Reader_Activation::get_client_id();
+		// Retrieve the email address associated with the client ID when the user was authenticated.
+		$email = get_transient( self::EMAIL_TRANSIENT_PREFIX . $client_id );
+		delete_transient( self::EMAIL_TRANSIENT_PREFIX . $client_id ); // Burn after reading.
+		if ( $email ) {
+			$existing_user = \get_user_by( 'email', $email );
+			if ( $existing_user ) {
+				// Log the user in.
+				$result = Reader_Activation::set_current_reader( $existing_user->ID );
+			} else {
+				$result = Reader_Activation::register_reader( $email );
+				// At this point the user will be logged in.
+			}
+			if ( is_wp_error( $result ) ) {
+				return $result;
+			}
+			Reader_Activation::save_current_user_login_method( 'google' );
+
+			return \rest_ensure_response(
+				[
+					'email'   => $email,
+					'message' => __(
+						'Thank you for registering!',
```

Comment:
Non-blocking suggestion: If the reader is logging in with a new session after having already registered with the same email in the past, can we show a "Thank you for signing in" message instead?

Reply:
Good idea, done in d1dba77ee7506393b7bf4a2b74f6225e8bf9af4d